015-kafka-topic-encryption.md

```diff
@@ -0,0 +1,194 @@
+<!-- This template is provided as an example with sections you may wish to comment on with respect to your proposal. Add or remove sections as required to best articulate the proposal. -->
+
+# Proxy-Based Kafka Per-Topic Encryption
+
+The goal of this proposal is to provide topic-level encryption-at-rest for Kafka such that distinct encryption keys can be used for different topics. 
+A core library is proposed which is deployed into a proxy. Proxies can flexibly be deployed in a variety of topologies and require no changes to Kafka clients and brokers. All encryption/decryption takes place at the proxy. An unencrypted topic's messages pass through the proxy with only the minimal overhead associated with proxying and message inspection.
+
+This proposal is based on a complete, working implementation.
+
+## Current situation
+
+Strimzi does not have a current capability in this area. 
+
+## Motivation
+
+Apache Kafka does not directly support any form of encryption-at-rest for data stored at a broker.
+Nevertheless, Kafka is increasingly used as a store of data, not just as a
+means of transferring it from one location to another. In an
+enterprise, this means that Kafka must conform to the same security and
+compliance requirements as conventional data storage systems such as relational databases. 
+To compensate, Kafka service providers typically use disk or file system encryption. 
+This approach has well-known shortcomings, most notably the ability of anyone with appropriate file system permissions, such as system administrators, to read data. Security best practices and compliance standards
+explicitly call for application- or database-level encryption to protect sensitive information against such exposures.
+This document proposes a technical solution for providing upper-layer encryption-at-rest for Kafka systems.
+
+## Proposal
+
+### Overview
+An implementation of topic encryption is proposed whereby the message stream between client and broker is intercepted. Incoming data messages (Produce requests) are inspected to determine whether their payload
+should be encrypted according to a policy. If so, the data portions are encrypted and the modified
+message is forwarded to the broker. As a result, the topic data is stored by the broker in encrypted form.
+On the reverse direction, encrypted responses (responses to Fetch requests) are decrypted prior to being sent to clients.
```

Comment:
@tombentley These are all important points you raise. 

- Agree, policy must be dynamically updated to, at the very least, accommodate topic creation and deletion. This should be done without restarting.  
- The proposal includes the KMS interface and at least one implementation of that interface initially, IBM Key Protect. We have started work with IBM Secrets Manager which is Vault. Very glad to discuss whether Vault would be a preferred initial KMS implementation.
- Migrating a key to another KMS can be supported however could require a restart, at least initially. The new KMS must of course be of a supported type. As long as the key material remains the same, migration is a configuration change.
- Yes, the credential is very sensitive data. It is used to access the KMS. Therefore the entire policy, in its current form,  must be treated as a secret.
- kms_instance is an optional parameter. To access IBM Key Protect, a URL and an instance ID are required. If a KMS does not require this parameter (e.g., Vault) it can be dropped. Perhaps there are other KMS systems out there which require an additional parameter in addition to a URL. kms_instance is a placeholder for such a param. 

I will provide an update with a brief description of the policy parameters.

Reply:
Thanks!

I think there's been most user interest (i.e. for TLS-related stuff) around Vault (@scholzj correct me if I'm wrong), so it would be great to have support for that.

I guess the KMS migration via restart should be fine (since we don't expect it to be a common thing). I hadn't realised that the way to do this would be to just export the old keys etc from the one to the other.

Reply:
KMS semantics should be the same as well when migrating. If AES data encryption keys are stored in the KMS, then migrating should be straightforward. However if an encryption model is employed which involves some sort of computation on the part of the KMS (e.g., envelope encryption with wrap/unwrap primitives) then one KMS is  presumably not a drop-in replacement since vendors may implement functions (e.g., envelope encryption) differently. The topic would then contain cryptographic artifacts from both KMS systems and could not be supported.  So migration needs to be qualified.

Reply:
@tombentley fyi I have updated the proposal to include clarification on policy updates.

Reply:
@chris-giblin could you fix the typo?